### ISSUE TITLE
support for boost for average for user defined loss functions

### ISSUE THREAD

**Issue by amirnachum15** [catboost/catboost#2443]
Hi!
Didn't already find a post about it - but is it possible to add support for boost_from_average=True in user defined loss functions?

this can greatly deter from using user defined functions, as in some scenarios performance will be impacted (when starting point will be very far away).
is it possible to add support to also define inside the loss function the boost_from_average behavior, or maybe even allow to fall back to the default MSE behavior in boost_from_average?

Thanks

**Comment by shubhamflip** [comment]
Any update on this ? As of now, we can't use boost_from_average=True with a use defined loss function and this significantly affect the performance of the model. Due to this reason I had to use other models like xgboost,lightgbm which allow this to be true even with custom loss function. Any timeline on when we can see this functionality in catboost ?

**Comment by nelsoncardenas** [comment]
I had the same issue; this would be a great and straightforward feature. People might find his custom cost functions performing worse than expected just because of this issue

**Comment by tiagofrepereira2012** [comment]
Indeed, that would be a nice feature to have.

**Comment by tiagofrepereira2012** [comment]
From the time being, a possible workaround until having this feature is to define an initial dummy model that predicts the average input.

For instance, in the example below, I train a model using one of the standard losses that allows you to use the boost_from_average feature. The model is set in a way to learn very little from the data, so we essentially have the average values as output.
This model is used as input to train the model with the customized loss.

```
import numpy as np
from catboost import CatBoostRegressor, MultiTargetCustomMetric, MultiTargetCustomObjective


class MultiRMSEObjective(MultiTargetCustomObjective):
    """
    Custom RMSE objective for multi-target regression.

    This is a copy from their documentation. Using this as a reference to create a
    custom objective for multi-target regression.
    """

    def calc_ders_multi(self, approx, target, weight):
        """
        Computes first derivative and Hessian matrix of the loss function
        with respect to the predicted value for each dimension.

        Parameters
        ----------
        approxes : indexed container of floats
            Predictions for each dimension of single object.

        targets : single expected value
            True label.

        weight : float, optional (default=None)
            Instance weight.

        Returns
        -------
            der1 : list-like object of float
            der2 : list of lists of float

        """

        w = weight if weight is not None else 1.0
        der1 = [(target[i] - approx[i]) * w for i in range(len(approx))]
        der2 = [-w for i in range(len(approx))]

        # Convert to numpy arrays, float
        der1 = np.array(der1, dtype=np.float64)
        der2 = np.array(der2, dtype=np.float64)

        return (der1, der2)


class MultiRMSEMetric(MultiTargetCustomMetric):
    def get_final_error(self, error, weight):
        return np.sqrt(error / (weight + 1e-38))

    def is_max_optimal(self):
        return False

    def evaluate(self, approxes, target, weight):
        assert len(target) == len(approxes)
        assert len(target[0]) == len(approxes[0])

        error_sum = 0.0
        weight_sum = 0.0

        for i in range(len(approxes[0])):
            w = 1.0 if weight is None else weight[i]
            weight_sum += w
            for d in range(len(approxes)):
                error_sum += w * ((approxes[d][i] - target[d][i]) ** 2)
        return error_sum, weight_sum


# Usage Example
# Create a sample dataset
X_train = np.random.rand(100, 10)
y_train = np.random.rand(100, 3)  # Assuming 3 target variables

# Setting up a dummy model that learns almost nothing
dummy_model = CatBoostRegressor(
    loss_function="MultiRMSE",
    leaf_estimation_method="Gradient",
    learning_rate=1e-6,
    boost_from_average=True,
    iterations=1,
).fit(X_train, y_train)


model_2 = CatBoostRegressor(
    loss_function=MultiRMSEObjective(),
    eval_metric=MultiRMSEMetric(),
    leaf_estimation_method="Gradient",
    bootstrap_type="Bayesian",
    # leaf_estimation_iterations=1,
)

model_2.fit(X_train, y_train, init_model=dummy_model)
```
